fix(compile): apply partition wrapper when loading AOT cached functions

devbyteai opened the pull request:
## Summary
  Fixes #31439 where VLLM_USE_AOT_COMPILE=1 causes 2x latency regression when use_inductor_graph_partition=True and cudagraph_mode=PIECEWISE.

  ## Root Cause

  The bug is in vllm/compilation/decorators.py. When loading AOT-compiled functions from cache, the code bypasses maybe_use_cudagraph_partition_wrapper at two locations:

  1. Line 374-376: Early return when aot_compiled_fn already set (subsequent calls)
  2. Line 431-438: Early return after loading from AOT cache (first call after restart)

  The maybe_use_cudagraph_partition_wrapper context manager calls torch._inductor.utils.set_customized_partition_wrappers() which is required at runtime for proper CUDA graph capture with inductor partitioning.

  ## Solution

  Wrap both AOT function call paths with the partition wrapper context to ensure partition wrappers are active at runtime.

  ## Impact

  1. AOT + partition + PIECEWISE: ~5s → ~2s (fixed)
  2. AOT without partition: ~2s → ~2s (no change)
  3. Non-AOT + partition: ~2s → ~2s (no change)

  No breaking changes. The fix only adds the missing context manager.

  ## Testing

  1. Added test_partition_wrapper_applied_on_aot_load in tests/compile/test_aot_compile.py
  2. Test verifies set_customized_partition_wrappers is called when loading from AOT cache
  3. Test requires PyTorch >= 2.10.0.dev
  4. Tested on 4x RTX 3090 setup

  ## Reproduction

  1. Run without AOT (fast ~2s):
     VLLM_USE_AOT_COMPILE=0 vllm bench latency --model meta-llama/Llama-3.2-1B -cc.use_inductor_graph_partition=True -cc.cudagraph_mode=PIECEWISE

  2. Run with AOT before fix (slow ~5s):
     VLLM_USE_AOT_COMPILE=1 vllm bench latency --model meta-llama/Llama-3.2-1B -cc.use_inductor_graph_partition=True -cc.cudagraph_mode=PIECEWISE

  3. Run with AOT after fix (fast ~2s):
     Same as #2, should now match non-AOT performance

devbyteai committed:
fix(compile): apply partition wrapper when loading AOT cached functions

Fixes #31439 where AOT compile caused 2x latency regression when
use_inductor_graph_partition=True and cudagraph_mode=PIECEWISE.

Root cause: The AOT load paths at decorators.py:374-379 and
decorators.py:434-440 bypassed maybe_use_cudagraph_partition_wrapper,
which sets up torch._inductor.utils.set_customized_partition_wrappers()
for proper CUDA graph capture with inductor partitioning.

Solution: Wrap AOT compiled function calls with partition wrapper
context to ensure partition wrappers are active at runtime.

Testing:
- Added test_partition_wrapper_applied_on_aot_load to verify partition
  wrappers are set/cleared when loading from AOT cache

Signed-off-by: Devbyteai <abud6673@gmail.com>

devbyteai committed:
test: cover both AOT code paths in partition wrapper test

Signed-off-by: Devbyteai <abud6673@gmail.com>

devbyteai commented on tests/compile/test_aot_compile.py:
```diff
@@ -148,6 +148,87 @@ def test_shape_env(monkeypatch: pytest.MonkeyPatch):
+            )
+            assert wrapper_calls[-1] is None, (
+                "Last call should clear the wrapper"
+            )
```
fixed

devbyteai committed:
style: fix lint issues (line length, unused import)

Signed-off-by: Devbyteai <abud6673@gmail.com>

ProExpertProg commented on tests/compile/test_aot_compile.py:
```diff
@@ -148,6 +148,110 @@ def test_shape_env(monkeypatch: pytest.MonkeyPatch):
+        args = (torch.randn(10, 10),)
+
+        with tempfile.TemporaryDirectory() as tmpdirname:
+            m.setenv("VLLM_CACHE_ROOT", tmpdirname)
```
Let's extract this into a pytest fixture `vllm_tmp_cache`, and use the builtin `tmp_path` pytest fixture

devbyteai replied:
Added vllm_tmp_cache fixture using tmp_path.

ProExpertProg reviewed (approved):
Looks good, just a few nits on cleaning up the tests

devbyteai committed:
refactor(test): use pytest fixtures and mocker for cleaner tests

Address code review feedback:
- Add vllm_tmp_cache fixture using tmp_path for temp cache directory
- Remove unnecessary monkeypatch.context() wrapper
- Use pytest mocker.spy() instead of custom mock implementation

Signed-off-by: Devbyteai <abud6673@gmail.com>

ProExpertProg commented on tests/compile/test_aot_compile.py:
```diff
@@ -24,6 +25,13 @@
+@pytest.fixture
+def vllm_tmp_cache(tmp_path: Path, monkeypatch: pytest.MonkeyPatch) -> Path:
+    """Fixture that sets VLLM_CACHE_ROOT to a temporary directory."""
+    monkeypatch.setenv("VLLM_CACHE_ROOT", str(tmp_path))
```
```suggestion
    monkeypatch.setenv("VLLM_CACHE_ROOT", str(tmp_path / "vllm_cache"))
```

devbyteai committed:
Update tests/compile/test_aot_compile.py

Co-authored-by: Luka Govedič <ProExpertProg@users.noreply.github.com>
Signed-off-by: DevByteAI <161969603+devbyteai@users.noreply.github.com>
